text/3323-restrictions.md

```diff
@@ -0,0 +1,465 @@
+scope is declared (such that it is just `pub`), then the item is visible everywhere. This behavior
+is preserved for `impl` and `mut`. When a restriction is used, the behavior is allowed only within
+the declared scope. While in most cases the default visibility is private, `pub` is default in some
+cases, namely `enum` variants, `enum` fields, and `trait` items. `impl` and `mut` will have a
+consistent default: when omitted entirely, the scope is inherited from `pub`. This is both what is
+most convenient and is what is required for backwards compatibility with existing code.
+
+When an `ImplRestriction` is present, implementations of the associated trait are only permitted
+within the designated path. Any implementation of the trait outside this scope is a compile error.
+When a `MutRestriction` is present, mutable uses of the associated field are only permitted within
+the designated path. Any mutable use of the field outside the scope is a compile error. Further, a
+`struct`, `union`, or `enum` variant containing fields with an associated `MutRestriction` may not
+be constructed with `struct` expressions unless all fields are unrestricted in the present scope.
+This is the case even if the field is not directly declared, such as when functional record updates
+are used.
+
+## "Mutable use" in the compiler
+
+The concept of a "mutable use" [already exists][mutating use method] within the compiler. This
+catches all situations that are relevant here, including `ptr::addr_of_mut!`, `&mut`, and direct
+assignment to a field, while excluding interior mutability. As such, formal semantics of what
+constitutes a "mutable use" are not stated here.
+
+[mutating use method]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/visit/enum.PlaceContext.html#method.is_mutating_use
+
+## Interaction with `trait` aliases
+
+Trait aliases cannot be implemented. As such, there is no concern about compatibility between the
+`impl` restriction and `trait` aliases.
+
+# Drawbacks
+
+- Additional syntax for macros to handle
+- More syntax to learn
+- While unambiguous to parse, `trait impl(crate) Foo` could be confusing due to its similarity to
+  `impl Foo`.
+
+# Alternatives
+
+- `impl` and `mut` restrictions could be attributes, similar to `#[non_exhaustive]`.
+  - The proposed syntax could by syntactic sugar for these attributes.
+- Visibility could be altered to accept restrictions as a type of parameter, such as
+  `pub(crate, mut = self)`. This is not ideal because restrictions are not permitted everywhere
+  visibility is. As a result, any errors would have to occur later in the compilation process than
+  they would be with the proposed syntax. It would also mean macro authors would be unable to accept
+  only syntax that would be valid in a given context. Further, some positions such as `enum`
+  variants do not semantically accept a visibility, while they do accept a restriction.
+
```

Comment:
```suggestion
- The current syntax prioritizes orthogonality, separating the `mut`/`impl` keyword from the scope restriction. This produces verbose syntax. Many users may want similar restrictions. Could we provide a simpler, nicer syntax if we provided less capabilities? Would a new keyword or two help? We could choose a syntax with more limited flexibility but less verbosity and more simplicity. For instance, `sealed` or `readonly`.

```

(Modulo word wrap.)

Not necessarily advocating for this, but it seems like an alternative worth serious consideration.

Reply:
To look at it a different way: suppose we were designing Rust syntax anew, knowing we wanted this feature. What would it look like? Probably like fields being readonly by default unless marked otherwise, and traits being sealed by default unless marked otherwise.